### ISSUE TITLE
Question on go.produce.channel.size and queue.buffering.max.messages

### ISSUE THREAD

**Issue by vdepatla** [confluentinc/confluent-kafka-go#194]
Description
===========

I am using channel based producer for producing messages to kafka, 

-> my create producer looks like this
func NewKafkaProducer() ProducerInterface {
	producerCfg := &kafka.ConfigMap{
		"bootstrap.servers": Configuration().KafkaBrokers,
		"default.topic.config": kafka.ConfigMap{"acks": RequiredAcks, "partitioner": Partitioner, "request.timeout.ms": Configuration().RequestTimeOut,
			"message.timeout.ms": Configuration().MessageTimeOut},
		"message.send.max.retries": Configuration().MessageMaxRetries,
		"retry.backoff.ms":         Configuration().MessageRetryBackOff,
		"go.produce.channel.size":  1,
               "queue.buffering.max.messages": 1,
	}
	kprod, err := kafka.NewProducer(producerCfg)
	panicIfError(err, "kafka.NewProducer()", "brokers", Configuration().KafkaBrokers)
	prod := KafkaProducer{producer: kprod}
	go prod.handleDelivery()
	return &prod
}

when creating a producer, I create a handle delivery go routine that can handle delivery reports which looks like below.

-> handle delivery
func (p *KafkaProducer) handleDelivery() {
	for {
		log.Printf("in handle delivery")
		for e := range p.producer.Events() {
			switch ev := e.(type) {
			case *kafka.Message:
				m := ev
				if m.TopicPartition.Error != nil {
					err = m.TopicPartition.Error
				}
				logAndInstrumentKafka(time.Now(), err, *m.TopicPartition.Topic, string(m.Key), m.TopicPartition.Offset.String(), m.Value)
			default:
				log.Printf("unknown event: %v", e)
			}
		}
	}
}
And the Produce() looks like below

-> produce
func (p *KafkaProducer) Produce(topic string, partitionKey string, content []byte) {
	tp := kafka.TopicPartition{Topic: &topic, Partition: kafka.PartitionAny}
	msg := &kafka.Message{TopicPartition: tp, Value: content, Key: []byte(partitionKey)}
	p.producer.ProduceChannel() <- msg
	log.Printf("len of producer channel %v", p.producer.Len())
}

Question: when kafka is down, and app keeps producing messages, does producechannel() keeps these in a queue so when kafka is up it can transmit ? how can I control how many messages the channel holds when kafka is down ?

I reduced the numbers for go.produce.channel.size  and queue.buffering.max.messages to 1 and stopped kafka on my local and produced around 100 messages expecting it to give error like "message queue full" but what I see is that all 100 messages gets published when I start the local kafka again.

How to reproduce
================

Checklist
=========
Please provide the following information:

 - [ ] confluent-kafka-go and librdkafka version (`LibraryVersion()`):
 - [1.10 ] Apache Kafka broker version:
 - [ ] Client configuration: `ConfigMap{...}`
 - [ ] Operating system:
 - [ ] Provide client logs (with `"debug": ".."` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by vdepatla** [comment]
what all parameters can be configured to make the internal queue full ?

**Comment by edenhill** [comment]
The Go client is a thin layer on top of the underlying librdkafka C library, which is the actual Kafka client implementation.

When producing the messages are stored in *librdkafka* internal queues for up to `message.timeout.ms` waiting for the partition to become available (broker leader available), handling temporary loss of connection, brokers, etc. This is so that the application does not need to be aware or handle these sort of problems.

This internal queue is limited by the `queue.buffering.max.messages` and `queue.buffering.max.kbytes`  configuration properties, when either one of those limits are exceeded the Produce() call will fail with `ErrQueueFull`.

The ProduceChannel is just a channel interface to the underlying Produce() call to give you blocking behaviour when the librdkafka queue is full.

**Comment by vdepatla** [comment]
Thank you.
1. How to know the length of internal queue ? is it from Len() ?
2. For testing, I have configured  "queue.buffering.max.messages": 1, "queue.buffering.max.kbytes": 1  but I never got ErrQueueFull.
with my local kafka down, I was able to write 1000000 messages to producer channel, I was not able to write more messages but didn't see an error either, Is there any other configuration that is taking precedence ?

**Comment by edenhill** [comment]
When using the ProducerChannel it will block when the queue limit is exceeded, until there is room on the queue.
If you were to call Produce() directly you would see ErrQueueFull after the first message.

**Comment by vdepatla** [comment]
I see .. I think that's what I see .. the channel is blocked after 1000000 (cannot write any more messages), if I start kafka at this time I see all messages sent to kafka .. if I don't then they timeout after 5 minutes

**Comment by vdepatla** [comment]
queue limit exceed seem to depend  go.produce.channel.size, queue.buffering.max.kbytes, queue.buffering.max.messages.

if go.produce.channel.size is higher, even though queue.buffering.max.messages, queue.buffering.max.kbytes are lower, the channel is not blocked. Is this correct ?

**Comment by edenhill** [comment]
Right, so librdkafka's producer queue limit is controlled by queue.buffering.max.*, when that one is full the Go ProducerChannel will begin to buffer messages, which is controlled by go.produce.channel.size.
Only when both queues are full will writing to the ProduceChannel block.

**Comment by vdepatla** [comment]
ok. by "ProduceChannel block" you mean the channel will wait for response ? if the time out exceeds while producer channel is blocked, it will unblock coz there will delivery reports for the messages with local message time out ?
